pkg/container/versioned/value.go

```diff
@@ -0,0 +1,510 @@
+	version atomic.Uint64
+}
+
+func (a *atomicVersion) load() version {
+	return version(a.version.Load())
+}
+
+func (a *atomicVersion) store(version version) {
+	a.version.Store(uint64(version))
+}
+
+type versionCount struct {
+	version version
+	count   int
+}
+
+// Coordinator defines a common version number space for multiple versioned.Values,
+// and provides facilities for cleaning out stale versions.
+// The Value's are not directly managed by the Coordinator, but all the values under coordination
+// should be cleaned by the 'cleaner' function given to the Coordinator.
+type Coordinator struct {
+	// Logger supplied to NewCoordinator. Should be set if logging is desired.
+	Logger *logrus.Entry
+
+	// Cleaner is called with the earliest version that must be kept.
+	// Must be set to clean up resources held for old versions.
+	// Cleaner function may be called concurrently, the function must synchronize
+	// use of any shared resources.
+	Cleaner func(KeepVersion)
+
+	// mutex protects the rest of the fields
+	mutex lock.RWMutex
+
+	// version is the version number of the last applied change
+	version version
+
+	// oldest version not cleaned off
+	oldestVersion version
+
+	// versions is an ordered list of outstanding VersionHandles with a reference count.
+	// Outdated values can be cleaned when there are no outstanding VersionHandles for them.
+	versions []versionCount
+}
+
+// PrepareNextVersion returns a transaction to be used when adding or removing values.
+//
+// Callers need to coordinate so that a single goroutine is performing modifications at any one
+// time, consisting of the following operations:
+//
+// - tx := coordinator.PrepareNextVersion()
+//   - value.SetAt(... , tx)
+//   - value.RemoveAt(..., tx)
+//   - ...
+//
+// - tx.Commit()
+func (v *Coordinator) PrepareNextVersion() *Tx {
+	v.mutex.Lock()
+	defer v.mutex.Unlock()
+	return &Tx{
+		nextVersion: v.version + 1,
+		coordinator: v,
+	}
+}
+
+// All values in Tx are constants
+type Tx struct {
+	nextVersion version
+	coordinator *Coordinator
+}
+
+// LatestTx refers to maxVersion without having a coordinator, should only be used for testing.
+var LatestTx = &Tx{nextVersion: maxVersion}
+
+func (tx *Tx) String() string {
+	return strconv.FormatUint(uint64(tx.nextVersion), 10)
+}
+
+func (tx *Tx) After(v KeepVersion) bool {
+	return tx.nextVersion > version(v)
+}
+
+// Commit makes a new version of values available to readers
+// Commit call may be omitted if no changes were actually made.
+func (tx *Tx) Commit() error {
+	return tx.coordinator.commit(tx.nextVersion)
+}
+
+// GetVersionHandle returns a VersionHandle for the given transaction.
+func (tx *Tx) GetVersionHandle() *VersionHandle {
+	// This is only needed to support LatestTx to make some testing easier
+	if tx.coordinator == nil && tx.nextVersion == maxVersion {
+		return Latest()
+	}
+	return tx.coordinator.getVersionHandle(tx.nextVersion)
+}
+
+func versionHandleCmp(a versionCount, b version) int {
```

Comment:
nit: I think this can be replaced with `cmp.Compare`

Reply:
wouldn't this require 'a' and 'b' to be of the same type, and `comparable`? Also, we do not want to take the `count' in `versionCount` into the comparison.

Reply:
err, must've not had my coffee when I wrote that >.< yep!